tests/unit/kernels_reference.cpp

```diff
@@ -174,3 +175,78 @@ std::vector<float> dw::reference::Transpose(const float* in, size_t rows, size_t
 
     return std::move(inT);
 }
+
+void dw::reference::CPUMaxPooling2DForward(const float* input, const deepworks::Shape& shape, float* output, const std::array<int, 2>& kernel,
+                                           const std::array<int, 2>& padding, const std::array<int, 2>& stride) {
+    const auto options = torch::TensorOptions()
+                         .dtype(torch::kFloat32)
+                         .layout(torch::kStrided);
+    const std::vector<int64_t> input_shape(shape.begin(), shape.end()); // IntArray doesn't copy data, it's wrapper around raw data
+    const torch::Tensor input_tensor = torch::from_blob(const_cast<void*>(reinterpret_cast<const void*>(input)), input_shape, options);
+
+    torch::Tensor result_tensor = torch::max_pool2d(input_tensor, {kernel[0], kernel[1]}, {stride[0], stride[1]}, {padding[0], padding[1]});
+    std::copy_n(result_tensor.data_ptr<float>(), result_tensor.numel(), output);
+}
+
+void dw::reference::CPUConvolution2DForward(const float* input, const deepworks::Shape& shape, const float* weights, const float* bias,
+                                            float* output, int out_channels, const std::array<int, 2>& kernel, const std::array<int, 2>& padding,
+                                            const std::array<int, 2>& stride) {
+    const auto options = torch::TensorOptions()
+                         .dtype(torch::kFloat32)
+                         .layout(torch::kStrided);
+    const int in_channels = shape[1];
+    const std::vector<int64_t> input_shape(shape.begin(), shape.end()); // IntArray doesn't copy data, it's wrapper around raw data
+    const std::vector<int64_t> weights_shape{out_channels, in_channels, kernel[0], kernel[1]};
+    const std::vector<int64_t> bias_shape{out_channels};
+
+    const torch::Tensor input_tensor = torch::from_blob(const_cast<void*>(reinterpret_cast<const void*>(input)), input_shape, options);
+    const torch::Tensor weights_tensor = torch::from_blob(const_cast<void*>(reinterpret_cast<const void*>(weights)), weights_shape, options);
+    const torch::Tensor bias_tensor = torch::from_blob(const_cast<void*>(reinterpret_cast<const void*>(bias)), bias_shape, options);
+
+    torch::Tensor result_tensor = torch::conv2d(input_tensor, weights_tensor, bias_tensor, {stride[0], stride[1]}, {padding[0], padding[1]});
+    std::copy_n(result_tensor.data_ptr<float>(), result_tensor.numel(), output);
+}
+
+void dw::reference::CPUConvolution2DBackward(const float* input, const float* grad_output, const deepworks::Shape& shape,
```

Comment:
`shape` it's hard to understand  what shape I should pass to the function

Reply:
because the API with raw pointer isn't descriptive at all